### ISSUE TITLE
Regarding the MRR and Hits metrics change with number of partitions

### ISSUE THREAD

**Issue by AnuradhaSK** [facebookresearch/PyTorch-BigGraph#94]
The following table contains the MRR and Hits values I got for the live journal dataset when it was trained with 0.001 learning rate and 5 epochs. Generally, these evaluation metrics results have increased when the number of partitions increases. I observed similar results for two other edge lists. However, the results given in the [paper](https://www.sysml.cc/doc/2019/71.pdf) for FB15k, freebase datasets have negligible changes when the number of partitions increases. Is there any specific reason to clarify this behaviour?

No of partitions | MRR | Hits@1 | Hits@10 | Hits@50
-- | -- | -- | -- | --
1 | 0.762561 | 0.705239 | 0.866899 | 0.955799
2 | 0.761561 | 0.704932 | 0.864738 | 0.954564
4 | 0.767408 | 0.710564 | 0.870976 | 0.957202
8 | 0.80069 | 0.745636 | 0.898649 | 0.966953
16 | 0.80881 | 0.753067 | 0.906994 | 0.969853
32 | 0.808281 | 0.751425 | 0.908283 | 0.970592

**Comment by lerks** [comment]
These are great questions but unfortunately I don't have an answer. My intuition has always been that using fewer partitions should achieve better results than using more. This is because, if you consider the limit of using only one partition, then the system is able to replicate any other partitioning scheme if it is lucky enough to shuffle the edgelist so that the edges of each bucket are consecutive. (What I am trying to say is that unpartitioned training is at least as expressive/powerful as any partitioning). Therefore I considered the FB15k case, where we don't "lose" anything by increasing the number of partitions, a lucky outcome, the best possible.

However, as you noticed and as other users also reported, it seems that experimentally the opposite phenomenon is observed: model quality is lower with fewer partitions. I still can't really explain why that happens. My current best guess is that it may be due to non-ideal hyperparameters: in theory, one should re-optimize the hyperparameters for each partitioning scheme. Maybe that will fix this apparent paradox...

In short, I think this is still an open question, where more research would be needed.

**Comment by adamlerer** [comment]
The MRR numbers reported by PBG are *intra-partition* MRR, i.e. negatives are only sampled from within the partition (because the whole point of partitioning is to avoid loading all the embeddings into memory).

For the paper, to compute exact MRR metrics we had to unpartition the embeddings and run evaluation on a single large partition. I suspect if you did that you'd see these effects go away.

In general, the *true* MRR should get worse as you increase the number of partitions (although by a very small amount as long as your partitions are reasonably large). As a best practice, you should set numPartitions as small as possible given the memory and parallelization constraints of your system.

**Comment by AnuradhaSK** [comment]
Thank you for the kind replies @lerks  and @adamlerer.
@adamlerer Can you please provide the steps to follow in order to unpartition the embeddings and run evaluation for live journal dataset?

**Comment by AnuradhaSK** [comment]
In order to get accuracy values for FB15k with different number of partitions, I changed  
 ```
entities={
            'all': {'num_partitions': 8},
        }
```
in the PyTorch-BigGraph/torchbiggraph/examples/configs/fb15k_config.py file and executed
`python3.6 -m torchbiggraph.examples.fb15k --config /home/anuradhak/anuradha/PyTorch-BigGraph/torchbiggraph/examples/configs/fb15k_config.py` command on the terminal. I encountered with the following error.

![PBG-Error](https://user-images.githubusercontent.com/25483865/64491491-af4cb000-d286-11e9-862a-a635c86a0b2f.png)

**Comment by lerks** [comment]
Yes, our implementation of filtered MRR (which you can find in `filtered_eval.py`) is quite rudimentary, because it is only used inside the FB15k demo. As such, it doesn't support partitioned entity types, and many other things (featurized entities, non-dynamic relations, non-all-negatives, ...). We don't have plans to improve that at the moment, because it's a lot of work to get it to scale well and we don't think it's a super useful metric compared to unfiltered MRR on large graphs (although this is debatable). However we'll gladly accept any pull request that improves it, if you want to take a stab at it!

**Comment by Nicolaus93** [comment]
Hi guys, I am using your library on some graphs with the number of nodes in the range 1M-10M, such as some versions of wikipedia that you can find here http://law.di.unimi.it/datasets.php
Since the whole datasets don't fit in memory for training, to learn the embeddings on my machine I split them in partitions. The training goes fine, the MRR and Hits metrics don't look bad. Furthermore, to check the quality of the embeddings, I'm using the k-means algo from FAISS. Once the algorithm has converged, I check the nodes closest to the centroids. If I do this on the embeddings from a single partition the results totally make sense, e.g. I have been able to observe clusters of football players, movies and songs, French writers and so on. However, when I try to run the algorithm on the whole set of embeddings, the results do not make sense anymore, i.e. clusters are not very well defined and I find pages and topics which really seem unrelated close to each other. 
Is this something I should expect or I am probably doing something wrong? Let me know if you need some info (i.e. about the code).

**Comment by lerks** [comment]
Here are a few thoughts:
- "Anecdotal" evaluation, by manually inspecting a handful of examples and subjectively judging the quality, is quite noisy and error prone. Also, I'm not sure whether "clustering using FAISS" is a sound evaluation method. Do you think you could find a more scientifically and statistically based method to evaluate?
- The fact that you find "good clusters" in individual partitions but don't find them when merging them suggests to me that there could be an error in how you do the merging. That is: you have a bunch of points (the embeddings of one partition) that are clustered nicely but when you just add more points to it (by merging them with another partition) you lose the clusters. This is odd (but not impossible).
- It may be possible that embeddings in different partitions aren't "aligned". This is a result of random initialization and training on disjoint buckets: each bucket converges to its own "alignment" independently of others and may thus disagree. We have however several measures in place to avoid this. First of all more training should fix it. Then, if you are in a single machine, using either the inside_out or outside_in bucket order will guarantee that buckets are only trained on after they have been (at least partly) initialized. If you're working in distributed mode (multiple machines), the same is achieved by the `distributed_tree_init_order` order. Both of these last settings are defaults, so you shouldn't have problems unless you explicitly changed them.

If the above doesn't help I guess we'll need to see your code and your config...

**Comment by AnuradhaSK** [comment]
Hi @adamlerer 
Can I know the steps to follow in order to unpartition the embeddings and run evaluation on a single large partition to get final MRR and Hit matrices?

**Comment by adamlerer** [comment]
@AnuradhaSK the code to unpartition the embeddings is not provided, and we didn't consider it a very viable approach because the whole reason to partition is because you can't fit it all in memory.

The way we did it for the paper is to pick a random "cache" of negatives (about 1M) that are used by all partitions, rather than the normal negatives from the partition. I think you'd have to reimplement that if you wanted to do it this way.

The other way, if it fits into memory, would be to export the final embeddings to TSV, then run import_to_tsv.py again with 1 partition, and then you'd have to write a function that took the embeddings TSV and the dictionary file and produced the embeddings_0 checkpoint to evaluate from (you could use the same global model checkpoint).

**Comment by adamlerer** [comment]
@Nicolaus93 we have not observed degradation in qualitative model quality when splitting into a (reasonable) number of partitions, although it's possible that for some graphs/settings that could happen, because each node is only being compared against negatives in its own partition. 1M nodes is pretty small so if you're splitting that into a lot of partitions, I wouldn't be surprised if there was some degradation. (1M nodes x 100dim = ~400MB, so I don't think you should *need* to partition it).

I'd be curious to see what parameters you used, as that might be responsible for the degradation.
